src/librustc_typeck/check/mod.rs

```diff
@@ -1214,55 +1213,12 @@ pub fn check_item_type<'a,'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, it: &'tcx hir::Item
 }
 
 fn check_on_unimplemented<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                                    def_id: DefId,
+                                    trait_def_id: DefId,
                                     item: &hir::Item) {
-    let generics = tcx.generics_of(def_id);
-    if let Some(ref attr) = item.attrs.iter().find(|a| {
-        a.check_name("rustc_on_unimplemented")
-    }) {
-        if let Some(istring) = attr.value_str() {
-            let istring = istring.as_str();
-            let name = tcx.item_name(def_id).as_str();
-            let parser = Parser::new(&istring);
-            let types = &generics.types;
-            for token in parser {
-                match token {
-                    Piece::String(_) => (), // Normal string, no need to check it
-                    Piece::NextArgument(a) => match a.position {
-                        // `{Self}` is allowed
-                        Position::ArgumentNamed(s) if s == "Self" => (),
-                        // `{ThisTraitsName}` is allowed
-                        Position::ArgumentNamed(s) if s == name => (),
-                        // So is `{A}` if A is a type parameter
-                        Position::ArgumentNamed(s) => match types.iter().find(|t| {
-                            t.name == s
-                        }) {
-                            Some(_) => (),
-                            None => {
-                                span_err!(tcx.sess, attr.span, E0230,
-                                                 "there is no type parameter \
-                                                          {} on trait {}",
-                                                           s, name);
-                            }
-                        },
-                        // `{:1}` and `{}` are not to be used
-                        Position::ArgumentIs(_) => {
-                            span_err!(tcx.sess, attr.span, E0231,
-                                                  "only named substitution \
-                                                   parameters are allowed");
-                        }
-                    }
-                }
-            }
-        } else {
-            struct_span_err!(
-                tcx.sess, attr.span, E0232,
-                "this attribute must have a value")
-                .span_label(attr.span, "attribute requires a value")
-                .note(&format!("eg `#[rustc_on_unimplemented = \"foo\"]`"))
-                .emit();
-        }
-    }
+    let item_def_id = tcx.hir.local_def_id(item.id);
+    // an error would be reported if this fails.
+    let _ = traits::OnUnimplementedDirective::of_item(
+        tcx, trait_def_id, item_def_id);
```

Comment:
Doesn't fit in <100?

Reply:
it does. this just went through a few revisions.